mentat/code_context.py

```diff
@@ -366,57 +322,117 @@ def _feature_relative_path(f: CodeFeature) -> str:
-        removed_paths = list[Path]()
-        for new_path in paths.keys():
-            if new_path in self.include_files:
-                removed_paths.append(new_path)
-                del self.include_files[new_path]
-        return removed_paths, invalid_paths
+    def include(self, path: Path, ignore_patterns: Iterable[Path | str] = []) -> Set[Path]:
+        """Add code to the context
+
+        '.' is replaced with '*' (recusively search the cwd)
+
+        Args:
+            `path`: can be a relative or absolute file path, file interval path, directory, or glob pattern.
+
+        Return:
+            A set of paths that have been successfully added to the context
+        """
+        ctx = SESSION_CONTEXT.get()
+
+        if str(path) == ".":
+            path = Path("*")
+
+        included_paths: Set[Path] = set()
+        try:
+            code_features = get_code_features_for_path(
+                path=path,
+                cwd=ctx.cwd,
+                ignore_patterns=[*ignore_patterns, *self.ignore_patterns, *ctx.config.file_exclude_glob_list],
+            )
+        except PathValidationException as e:
+            ctx.stream.send(e, color="light_red")
+            return included_paths
+
+        for code_feature in code_features:
+            self.include_files[code_feature.path] = code_feature
+            included_paths.add(code_feature.path)
+
+        return included_paths
+
+    def exclude(self, path_: Path) -> Set[Path]:
+        """Remove code from the context
+
+        Args:
+            `path`: can be a relative or absolute file path, file interval path, directory, or glob pattern.
+
+        Return:
+            A set of paths that have been successfully removed from the context
+        """
+        ctx = SESSION_CONTEXT.get()
+
+        validated_path = validate_and_format_path(path_, ctx.cwd, check_for_text=False)
+
+        excluded_paths: Set[Path] = set()
+        try:
+            # file
+            if validated_path.is_file():
+                if validated_path not in self.include_files:
+                    ctx.stream.send(f"Path {validated_path} not in context", color="light_red")
+                    return excluded_paths
+                excluded_paths.add(validated_path)
+                del self.include_files[validated_path]
+            # file interval
+            elif ":" in str(validated_path):
+                _interval_path, _ = str(validated_path).split(":", 1)
+                interval_path = Path(_interval_path)
+                if interval_path not in self.include_files:
+                    ctx.stream.send(f"Path interval {validated_path} not in context", color="light_red")
+                    return excluded_paths
+                excluded_paths.add(interval_path)
+                del self.include_files[interval_path]
+            # directory
+            elif validated_path.is_dir():
+                if validated_path not in self.include_files:
```

Comment:
I don't think directories are ever in self.include_files; the point of self.include_files is that it has all of the individual files; if a user puts a directory in, they want all the children of the directory that are included to be excluded

Reply:
Yup that's correct. Will fix this